token/js/test/unit/transferHook.test.ts

```diff
@@ -1,173 +1,547 @@
+                keys: [
+                    { pubkey: sourcePubkey, isSigner: false, isWritable: true },
+                    { pubkey: mintPubkey, isSigner: false, isWritable: false },
+                    { pubkey: destinationPubkey, isSigner: false, isWritable: true },
+                    { pubkey: authorityPubkey, isSigner: true, isWritable: false },
+                ],
+                programId: transferHookProgramId,
+            });
+
+            const hydratedInstruction = await addExtraAccountMetasForExecute(
+                connection,
+                rawInstruction,
+                transferHookProgramId,
+                sourcePubkey,
+                mintPubkey,
+                destinationPubkey,
+                authorityPubkey,
+                amount,
+            );
+
+            const checkMetas = [
+                { pubkey: sourcePubkey, isSigner: false, isWritable: true },
+                { pubkey: mintPubkey, isSigner: false, isWritable: false },
+                { pubkey: destinationPubkey, isSigner: false, isWritable: true },
+                { pubkey: authorityPubkey, isSigner: true, isWritable: false },
+                { pubkey: extraMeta1Pubkey, isSigner: false, isWritable: false },
+                { pubkey: extraMeta2Pubkey, isSigner: false, isWritable: false },
+                { pubkey: extraMeta3Pubkey, isSigner: false, isWritable: false },
+                { pubkey: extraMeta4Pubkey, isSigner: false, isWritable: false },
+                { pubkey: extraMeta5Pubkey, isSigner: false, isWritable: false },
+                { pubkey: extraMeta6Pubkey, isSigner: false, isWritable: false },
+                { pubkey: transferHookProgramId, isSigner: false, isWritable: false },
+                { pubkey: validateStatePubkey, isSigner: false, isWritable: false },
+            ];
+
+            expect(hydratedInstruction.keys).to.eql(checkMetas);
+        });
+
+        it('can create a transfer instruction with extra metas', async () => {
+            const transferHookProgramId = Keypair.generate().publicKey;
+
+            const sourcePubkey = Keypair.generate().publicKey;
+            const mintPubkey = Keypair.generate().publicKey;
+            const destinationPubkey = Keypair.generate().publicKey;
+            const authorityPubkey = Keypair.generate().publicKey;
+            const validateStatePubkey = getExtraAccountMetaAddress(mintPubkey, transferHookProgramId);
+            const amount = 100n;
+            const decimals = 0;
+
+            // Arbitrary program ID included to test external PDAs
+            const arbitraryProgramId = Keypair.generate().publicKey;
+
+            const connection = await getConnection();
+            connection.getAccountInfo = async function mockFetchAccountDataFn(
+                publicKey: PublicKey,
+                _commitmentOrConfig?: Parameters<Connection['getAccountInfo']>[1]
+            ): ReturnType<Connection['getAccountInfo']> {
```

Comment:
nit: this function is pretty big, and very similar to the one in the previous test -- you could initialize all of the keys one level up and move this function up one level to be re-used by both tests, which would probably make this test more legible too.